docs/design-proposals/2023-11 TraceQL Extensions.md

```diff
@@ -0,0 +1,152 @@
+---
+Authors: Martin Disibio (@mdisibio), Joe Elliott (@joe-elliott), Jennie Pham (@ie-pham), Adrian Stoewer (@stoewer)
+Created: 2023 July - 2023 November
+Last updated: 2023-11-28
+---
+
+# TraceQL Extensions
+
+## Summary
+
+This design document describes extensions to the existing TraceQL language that allow for searching new fields and data types.
+
+This document is **not** meant to be a complete language specification. Rather it is meant to invite the community to comment on the basic structure and concepts of the language extensions.
+
+### Table of Contents
+
+- [Escaping Attribute Names](#escaping-attribute-names)
+- [Additional Scopes](#additional-scopes)
+  - [Trace](#trace)
+  - [Instrumentation Scope](#instrumentation-scope)
+  - [Event](#event)
+  - [Link](#link)
+- [Scoped Intrinsics](#scoped-intrinsics)
+- [Additional Data Types](#additional-data-types)
+  - [Arrays](#arrays)
+  - [IDs](#ids)
+- [Future Considerations](#future-considerations)
+
+## Escaping Attribute Names
+
+OpenTelemetry supports [any valid Unicode sequence](https://opentelemetry.io/docs/specs/otel/common/attribute-naming/) for attribute names. This includes characters problematic for a query language such as whitespace, mathematical symbols, various braces, etc. Previously TraceQL did not support [these problematic characters](https://github.com/grafana/tempo/blob/c00e7ef1103a8e7804b6f71fe8e90d0a2d9a9b02/pkg/traceql/lexer.go#L344) in attributes names.
+
+TraceQL will be extended to support these characters by allowing attribute names wrapped in `"` to be unparsed.
+
+`{ span."attribute with spaces" = "foo"}`
+
+TraceQL will also support the escape sequences `\\` and `\"` in an attribute name.
+
+`{ span."this is \"bad\"" = "foo"}`
```

Comment:
How about adding support for single quotes too, so users don't have to deal with escaping. For reference there is a plan to do that in prometheus to address full utf-8 support. Proposal [link](https://github.com/prometheus/proposals/blob/main/proposals/2023-08-21-utf8.md#syntax-examples).

Reply:
thanks for the heads up. i like the way they've used both `"` and `'` for their quoted attributes. it's something we should definitely keep in mind for a future improvement.